### PR TITLE
Fix causal cross-attention alignment in Triton backward

### PR DESCRIPTION
**Pull request by huiyeruzhou**
## Summary

- align generic Triton backward causal masking with the forward kernel for `seqlen_q != seqlen_k`
- apply the bottom-right causal offset to dK/dV and dQ loop bounds and masks
- add a causal cross-attention GQA regression test against PyTorch math SDPA

## Problem

The forward path uses bottom-right causal alignment (`q + seqlen_k - seqlen_q >= k`), while the generic backward path used square self-attention alignment (`q >= k`). For causal cross-attention with `seqlen_q < seqlen_k`, the forward output was correct but backward gradients could be incorrect or NaN.

This problem can be triggered in DiffusionGemma, where ffpa is used the efficient attention kernel for Gemma 4 and the diffusion procedure requires cross-attention.

## Validation

Validated on NVIDIA A100-SXM4-80GB with PyTorch 2.9.1 and Triton 3.5.1:

- targeted fp16/bf16 regression: 2 passed
- independent comparison against PyTorch math SDPA: relative L2 errors approximately 4e-4 to 6.5e-4 for fp16 and 3.6e-3 to 5.2e-3 for bf16
- forward and dQ/dK/dV are finite and aligned with the reference

### CODE REVIEW

**Comment by huiyeruzhou** [comment]
## Minimal reproduction

Run this on the unmodified upstream `main` and then on this PR:

```python
import math

import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel

from ffpa_attn import ffpa_attn_func
from ffpa_attn.functional import TritonBackend

B, Hq, Hkv, Nq, Nkv, D = 1, 8, 2, 512, 1024, 320
dtype = torch.float16
torch.manual_seed(11)

q = torch.randn(B, Hq, Nq, D, device="cuda", dtype=dtype, requires_grad=True)
k = torch.randn(B, Hkv, Nkv, D, device="cuda", dtype=dtype, requires_grad=True)
v = torch.randn(B, Hkv, Nkv, D, device="cuda", dtype=dtype, requires_grad=True)
grad_out = torch.randn_like(q)
scale = 1.0 / math.sqrt(D)

out = ffpa_attn_func(
    q,
    k,
    v,
    is_causal=True,
    scale=scale,
    enable_gqa=True,
    forward_backend=TritonBackend(forward=True),
    backward_backend=TritonBackend(backward=True),
)
out.backward(grad_out)

q_ref = q.detach().clone().requires_grad_(True)
k_ref = k.detach().clone().requires_grad_(True)
v_ref = v.detach().clone().requires_grad_(True)
group_size = Hq // Hkv
k_expanded = k_ref.repeat_interleave(group_size, dim=1)
v_expanded = v_ref.repeat_interleave(group_size, dim=1)

# FFPA forward uses bottom-right causal alignment for non-square attention:
# query row m can attend through key m + (Nkv - Nq).
row = torch.arange(Nq, device="cuda")[:, None]
col = torch.arange(Nkv, device="cuda")[None, :]
causal_mask = col <= row + (Nkv - Nq)
with sdpa_kernel(SDPBackend.MATH):
    out_ref = F.scaled_dot_product_attention(
        q_ref, k_expanded, v_expanded, attn_mask=causal_mask, scale=scale
    )
out_ref.backward(grad_out)

rel_l2 = torch.linalg.vector_norm((q.grad - q_ref.grad).float()) / torch.linalg.vector_norm(q_ref.grad.float())
print(f"dQ relative L2 error: {rel_l2.item():.6f}")
torch.testing.assert_close(q.grad, q_ref.grad, atol=1e-2, rtol=1e-2)
torch.testing.assert_close(k.grad, k_ref.grad, atol=1e-2, rtol=1e-2)
torch.testing.assert_close(v.grad, v_ref.grad, atol=1e-2, rtol=1e-2)
```

Observed on NVIDIA A100-SXM4-80GB:

- upstream `main` (`19f04db`): fails; 82.3% of dQ elements are outside tolerance, max absolute error 0.6167
- this PR: passes for both fp16 and bf16

The forward path and reference both use `q + (Nkv - Nq) >= k`; the original generic Triton backward incorrectly used `q >= k`.

**Review by DefTruth — APPROVED** [review]
